### PR TITLE
Add an method for search spans within a trace

### PR DESCRIPTION
**Pull request by B-Step62**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/B-Step62/mlflow/pull/13984?quickstart=1)

#### Install mlflow from this PR

```
# Use `%sh` to run this command on Databricks
OPTIONS=$(if pip freeze | grep -q 'mlflow @ git+https://github.com/mlflow/mlflow.git'; then echo '--force-reinstall --no-deps'; fi)
pip install $OPTIONS git+https://github.com/mlflow/mlflow.git@refs/pull/13984/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 13984
```

</p>
</details>

### What changes are proposed in this pull request?

Add `Trace.search_spans` method to search for spans within a single trace with several criteria. This is useful in some scenario, for example, when user wants to extract particular tool span for evaluation.

```
trace.search_spans(name="my_tool")
```

### How is this PR tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [x] Manual tests
- [ ] 
Screenshot 2024-12-05 at 10 14 38

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [x] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

Add `Trace.search_spans()` API to find particular spans with several search criteria (name, type).

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [x] No (this PR will be included in the next minor release)

### CODE REVIEW

**Review by daniellok-db — APPROVED** [review]
looks great, thanks!

File: mlflow/entities/trace.py [hunk 1]
```diff
@@ -100,6 +105,88 @@ def to_pandas_dataframe_row(self) -> dict[str, Any]:
             "tags": self.info.tags,
         }
 
+    def search_spans(
+        self, span_type: Optional[SpanType] = None, name: Optional[Union[str, Pattern]] = None
```
Comment (harupy) [inline mlflow/entities/trace.py:109]:
```suggestion
        self, span_type: Optional[SpanType] = None, name: Optional[Union[str, re.Pattern]] = None
```

https://docs.python.org/3/library/typing.html#typing.Pattern is deprecated.

File: mlflow/entities/trace.py [hunk 2]
```diff
@@ -100,6 +105,88 @@ def to_pandas_dataframe_row(self) -> dict[str, Any]:
...
+            y = run(2)
+            trace = mlflow.get_last_active_trace()
+
+            # 1. Search spans by name (exact match)
+            spans = trace.search_spans(name="add_one")
+            print(spans)
+            # Output: [Span(name='add_one', ...)]
+
+            # 2. Search spans by name (regular expression)
+            pattern = re.compile("add.*")
```
Comment (harupy) [inline mlflow/entities/trace.py:162]:
```suggestion
            pattern = re.compile(r"add.*")
```

File: mlflow/entities/trace.py [hunk 3]
```diff
@@ -100,6 +105,88 @@ def to_pandas_dataframe_row(self) -> dict[str, Any]:
...
+            print(spans)
+            # Output: [Span(name='add_one', ...)]
+        """
+
+        def _match_name(span: Span) -> bool:
+            if isinstance(name, str):
+                return span.name == name
+            elif isinstance(name, re.Pattern):
+                return name.match(span.name) is not None
+            return True
```
Comment (harupy) [inline mlflow/entities/trace.py:183]: What shoud if a user passes an int to name by mistake?

File: docs/source/llms/tracing/index.rst [hunk 4]
```diff
@@ -1502,3 +1502,70 @@ Alternatively, you can use the :py:func:`mlflow.search_traces` function to get t
...
+    print(spans)
+    # Output: [Span(name='add_one', ...)]
+
+    # Search for a span with the span type "TOOL"
+    spans = trace.search_spans(span_type=SpanType.TOOL)
+    print(spans)
+    # Output: [Span(name='add_one', ...), Span(name='add_two', ...), Span(name='multiply_by_two', ...)]
+
+    # Search for spans whose name starts with "add"
+    spans = trace.search_spans(name=re.compile("add.*"))
```
Comment (harupy) [inline docs/source/llms/tracing/index.rst:1569]:
```suggestion
    spans = trace.search_spans(name=re.compile("^add.*"))
```

nit

File: mlflow/entities/trace.py [hunk 5]
```diff
@@ -100,6 +105,88 @@ def to_pandas_dataframe_row(self) -> dict[str, Any]:
...
+            spans = trace.search_spans(name="add_one", span_type=SpanType.TOOL)
+            print(spans)
+            # Output: [Span(name='add_one', ...)]
+        """
+
+        def _match_name(span: Span) -> bool:
+            if isinstance(name, str):
+                return span.name == name
+            elif isinstance(name, re.Pattern):
+                return name.match(span.name) is not None
```
Comment (harupy) [inline mlflow/entities/trace.py:182]:
```suggestion
                return name.search(span.name) is not None
```

should we use `search` here?

File: tests/entities/test_trace.py [hunk 6]
```diff
@@ -229,3 +237,43 @@ def test_trace_pandas_dataframe_columns():
...
     assert Trace.pandas_dataframe_columns() == list(t.to_pandas_dataframe_row())
+
+
+@pytest.mark.parametrize(
+    ("span_type", "name", "expected"),
+    [
+        (None, None, ["run", "add_one_1", "add_one_2", "add_two", "multiply_by_two"]),
+        (SpanType.CHAIN, None, ["run"]),
+        (None, "add_two", ["add_two"]),
+        (None, re.compile("add.*"), ["add_one_1", "add_one_2", "add_two"]),
```
Comment (harupy) [inline tests/entities/test_trace.py:248]:
```suggestion
        (None, re.compile(r"add.*"), ["add_one_1", "add_one_2", "add_two"]),
        (None, re.compile(r"_two$"), ["add_two"]),
```

Let's add one more test case for regular expression

File: mlflow/entities/trace.py (hunk 3 above)
Comment (B-Step62) [reply mlflow/entities/trace.py:183]: Good callout, we should raise an error for such case.

File: mlflow/entities/trace.py (hunk 5 above)
Comment (B-Step62) [reply mlflow/entities/trace.py:182]: Ah true, great catch!

**Review by harupy — APPROVED** [review]
LGTM!

**Review by TomeHirata — APPROVED** [review]
LGTM

**Pull request merged**